Update docs for signature labels (#1396)
project-oak/oak: docs/concepts.md

    connection itself; this represents the actual authority of the caller.
  - an empty integrity component.

### Signature labels

Oak Wasm modules can be signed using [Ed25519](https://ed25519.cr.yp.to/)
scheme. Such signatures may be created by an Oak module reviewer and published

> on a public resource for Oak clients to download.

Author: Yes, clients download public keys, updated

In case when a client trusts a certain reviewer, it can use a corresponding

> author's public key and use it as a data label. This will allow an Oak server to

Author: Done

download and use new versions of modules signed by the same reviewer without
client's notice, i.e. client wouldn't need to constantly update module hashes in
data labels. It will also allow new versions of Oak modules to access old
client's data saved by Oak.

In order to clearer understand how Oak module signatures are used, here is a
possible life cycle of an Oak application: from development to usage.

#### Oak signature life cycle

The parties involved in the signature life cycle are:

> - Application developer

Author: I tried to merge these two lists, but if feels a little bit odd, since the second list should be a life cycle (too much text will be in one place)

  - Person/organization who writes an Oak application
- Oak module developer
  - Person/organization who develops and publishes Oak Wasm modules to be used
    by the **Application developer**
- Reviewer
  - Person/organization who reviews and signs Oak modules

> - Maintainer

Author: Changed to operator

    connection itself; this represents the actual authority of the caller.
  - an empty integrity component.

### Signature labels

Oak Wasm modules can be signed using [Ed25519](https://ed25519.cr.yp.to/)
scheme. Such signatures may be created by an Oak module reviewer who then

> publish corresponding public keys on a public resource. By signing a module

Reviewer: ```suggestion
publishes corresponding public keys on a public resource. By signing a module
```

Author: Done

> publish corresponding public keys on a public resource. By signing a module

Reviewer: The main thing is that they publish the signatures. Then their public key is also assumed to be published already.

Author: Updated `publishes signatures and corresponding public keys`

signing a module reviewers endorse it's functionality and show that they trust
this particular module not to have any undocumented functionality. The same

> reviewer may also use distinct keys for different purpose, for example to use a

Reviewer 2: ```suggestion
reviewer may also use distinct keys for different purposes, for example to use a
```

separate public key to sign a family of related modules.

> In case when a client trusts a certain reviewer, it can download a corresponding

Reviewer 3: I don't think this will be a typical use case. The main use case would be where the end-user trusts the application developer, and the application developer ensures that the client application attaches the correct public key label to the data to ensure that it could be processed in future (e.g. after being stored) even if code must be updated (e.g. to fix a bug).

Author: Done
Added:
```
Reviewer's public key can also be assigned to outgoing requests by the client library, which was created by the
Oak application developer. In this case client doesn't need to manually download
a public key, but they will need to trust that the client library assigns these public keys to outgoing requests correctly.
```

reviewer's public key and use it as a data label. This will allow an Oak server
to download and use new versions of modules signed by the same reviewer without
client's notice, i.e. client wouldn't need to constantly update module hashes in
data labels. It will also allow new versions of Oak modules to access old
client's data saved by Oak.

> In order to clearer understand how Oak module signatures are used, here is a

Reviewer: ```suggestion
In order to more clearly show how Oak module signatures are used, here is a
```

Author: Done

#### Oak signature life cycle

The parties involved in the signature life cycle are:

- Application developer
  - Person/organization who writes an Oak application
- Oak module developer
  - Person/organization who develops and publishes Oak Wasm modules to be used
    by the **Application developer**
- Reviewer
  - Person/organization who reviews and signs Oak modules
- Operator
  - Person/organization who deploys Oak applications on a production server
- Client
  - Person/organization who sends data to a deployed Oak application

The cycle itself looks as follows:

- **Reviewer**
  - Reviews existing Oak modules and checks their behavior
  - Signs Oak modules and publishes corresponding signatures (as well as public
    keys used for signing)
- **Application developer**:
  - Chooses which third-party Oak modules to use
  - Creates an Oak application using these third-party Oak modules
  - Publishes the application
- **Operator**
  - Downloads an Oak application
  - Downloads files containing Oak modules' signatures/keys
  - Deploys the Oak application and passes it a list of Oak module
    signatures/keys
- **Client**
  - Chooses which Oak modules' reviewer to trust

>   - Downloads corresponding reviewers' public keys

Reviewer 3: This does not seem to be a realistic scenario. Only a very small number of users would feel comfortable with donwloading a public key and ensuring that the appropriate public key abel is attached to all data.

Reviewer 2: Is it uncomfortable because users do not trust this approach or because it is too much work?

Reviewer 3: I think it would be confusing for most users.

Author: Updated
